### PR TITLE
(PCP-251) support for saving stdout/stderr/exitcode to files

### PR DESCRIPTION
**Pull request by mruzicka**
The pxp-module-puppet's run action can now save its standard output / standard error / exit code to files specified in the action's input.

### CODE REVIEW

File: modules/pxp-module-puppet [hunk 1]
```diff
@@ -235,13 +284,14 @@ if __FILE__ == $0
   if action == 'metadata'
     puts metadata.to_json
   else
-    params_and_config = JSON.parse($stdin.read.chomp) rescue nil
+    args = JSON.parse($stdin.read.chomp) rescue nil
 
-    run_result = run(params_and_config)
-    puts run_result.to_json
+    # TODO(ale): retrieve action_output instead of action_results,
```
Comment (parisiale) [inline modules/pxp-module-puppet]: This can be removed; it's fine as it is.

File: modules/pxp-module-puppet [hunk 2]
```diff
@@ -124,11 +125,12 @@ def start_run(config, params)
   end
 
   exitcode = run_result.exitstatus
+  output = run_result.to_s
```
Comment (parisiale) [inline modules/pxp-module-puppet]: Do you mind calling this one `puppet_output`?

**Comment by parisiale** [comment]
:+1: 
Tested with https://github.com/puppetlabs/pxp-agent/pull/305; we're able to restart pxp-agent in the middle of a run without losing the results.

Perhaps you could squash the last three commits once were done with this.
Make sure there's an empty "new line" at the end of the files :-P

@richardc please have a look at the changes.

**Comment by mruzicka** [comment]
I've force pushed this PR with minor changes discussed with @parisiale in HipChat.
AFAIKT all the files modified in the frame of this PR end in a new line character:
```
[mruzicka@michalruzickambp pxp-agent]$ od -ta modules/pxp-module-puppet | tail -2
0017440   sp   e   n   d  nl   e   n   d  nl                            
0017451
[mruzicka@michalruzickambp pxp-agent]$ od -ta modules/spec/unit/modules/puppet_spec.rb | tail -2
0037500   sp  sp   e   n   d  nl   e   n   d  nl                        
0037512
[mruzicka@michalruzickambp pxp-agent]$ git status
On branch PCP-251_simple_output_files_support
Your branch is up-to-date with 'origin/PCP-251_simple_output_files_support'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	input.json
	pxp-agent.conf

nothing added to commit but untracked files present (use "git add" to track)
```

I'm happy to squash the commits into one once this passes review.

**Comment by parisiale** [comment]
I've closed the PR by mistake.

File: modules/pxp-module-puppet [hunk 3]
```diff
@@ -221,12 +225,57 @@ def run(params_and_config)
...
+      else
+        1
+      end
+
+      begin
+        File.open(output_files["exitcode"], 'w') do |f|
+          f.print(status)
+        end
+      rescue
+        exit 5 # this exit code is reserved for problems with opening
```
Comment (parisiale) [inline modules/pxp-module-puppet]: You should do as above (print make_error_result()).

Comment (mruzicka) [inline modules/pxp-module-puppet]: At this point the result should have been written - mind you that this is an ```at_exit``` block. The problem is only with creating/writing the ```exitcode``` file.

Comment (parisiale) [inline modules/pxp-module-puppet]: It doesn't matter.

If pxp-agent is running, it will consider this a failure because it will get `exitcode == 5`.
If pxp-agent is not running, once it restarts, it will consider this as a failure, since the "exitcode" file is not in the results directory.
This will change anyway when we deal with exception handling; the module has to report its own errors on `stderr` or on the "stderr" file, not on the results object.
Also, consider that this is an edge case; pxp-module-puppet would need to successfully write to "stdout" and "stderr" but fail to write on "exitcode".

File: modules/pxp-module-puppet [hunk 4]
```diff
@@ -91,12 +95,9 @@ def get_result_from_report(exitcode, config, start_time)
...
   end
 
-  run_result = last_run_result(exitcode)
-
-  if exitcode != 0
-    run_result = make_error_result(exitcode, Errors::NonZeroExit,
-                                   "Puppet agent exited with a non 0 exitcode")
-  end
+  run_result = exitcode == 0 ?  last_run_result(exitcode)
+    : make_error_result(exitcode, Errors::NonZeroExit,
```
Comment (richardc) [inline modules/pxp-module-puppet]: Use of the ternary here over multiple lines is really hard to read.  If we're going to touch this for clarity I'd suggest a more explicit structure.

```
if exitcode == 0
  run_result = last_run_result(exitcode)
else
  run_result = make_error_result(exitcode, Errors::NonZeroExit,  "Puppet agent exited with a non 0 exitcode")
end
```

File: modules/pxp-module-puppet [hunk 5]
```diff
@@ -109,10 +110,10 @@ def get_result_from_report(exitcode, config, start_time)
   return run_result
 end
 
-def start_run(config, params)
+def start_run(config, action_input)
```
Comment (richardc) [inline modules/pxp-module-puppet]: If it's appropriate to rename `params` -> `action_input` here, isn't it also appropriate to rename them at the point of use?  https://github.com/mruzicka/pxp-agent/blob/PCP-251_simple_output_files_support/modules/pxp-module-puppet#L64

I don't think the rename adds much, but if done it should be consistent.

File: modules/pxp-module-puppet [hunk 6]
```diff
@@ -154,7 +156,7 @@ def metadata()
           },
           :required => [:env, :flags]
         },
-        :output => {
+        :results => {
```
Comment (richardc) [inline modules/pxp-module-puppet]: This seems unrelated to the change at hand.

Comment (mruzicka) [inline modules/pxp-module-puppet]: @parisiale Do you depend on this in your changes?

File: modules/pxp-module-puppet [hunk 7]
```diff
@@ -235,13 +286,11 @@ if __FILE__ == $0
...
 
-    run_result = run(params_and_config)
-    puts run_result.to_json
+    action_results = action_run(args)
+    print action_results.to_json
 
-    if !run_result["error"].nil?
-      exit 1
-    end
+    exit 1 unless action_results["error"].nil?
```
Comment (richardc) [inline modules/pxp-module-puppet]: Preferred the original expression of this (explicit if vs adverbial unless).  Adverbial unless shifts the eye to the action, but I think when following this code the test is more important to encounter first.

File: modules/pxp-module-puppet (hunk 6 above)
Comment (parisiale) [inline modules/pxp-module-puppet]: PCP-251 is about both that schema change and how to convey the output.

I think the point here is whether or not this should go in a separate commit or PR. I don't have any strong opinion on that.

File: modules/pxp-module-puppet [hunk 8]
```diff
@@ -221,12 +225,59 @@ def run(params_and_config)
...
+end
+
+def action_run(args)
+  if !args
+    return make_error_result(DEFAULT_EXITCODE, Errors::InvalidJson,
+                             "Invalid json parsed on STDIN. Cannot start run action")
   end
 
-  return start_run(config, params)
+  output_files = args["output_files"]
```
Comment (richardc) [inline modules/pxp-module-puppet]: Oh, this seems like a terrible place to pull this from.

**Comment by richardc** [comment]
Hang on.  From the actions *input*?  As in the paths will be specifiable in the pxp command message?  That seems like an odd place to put the responsibility of picking a sane location for a spool directory.

Also if it is there, then shouldn't the action's input schema also change to reflect this?

**Comment by parisiale** [comment]
I think you missed a lot of what was going on here.

For Ankeny, pxp-agent was passing PXP request params + config to the module. And the module was required to write its output on stdout.

Now that's formalized.
On top of that, pxp-agent may put the module in charge of writing its output to file. That is done via the optional "output_files" entry. In that case the module is required to write output on file.

File: modules/pxp-module-puppet (hunk 6 above)
Comment (richardc) [inline modules/pxp-module-puppet]: Well this is changing output (from the old way of running a module?) to results (from the new way I assume) so I'd expect this change to be part of the 'port to the new api' change, rather than the 'add this new behaviour' change.

**Comment by richardc** [comment]
OK, feels weird but I see the code does that.

Generally seems it does what it should, but there are style nits I observed.

**Comment by mruzicka** [comment]
I've added another commit addressing the style issues.
@richardc please ping me if you are happy with it.

File: modules/pxp-module-puppet [hunk 9]
```diff
@@ -221,12 +228,59 @@ def run(params_and_config)
...
+  if output_files
+    begin
+      $stdout.reopen(File.open(output_files["stdout"], 'w'))
+      $stderr.reopen(File.open(output_files["stderr"], 'w'))
+    rescue => e
+      print make_error_result(DEFAULT_EXITCODE, Errors::InvalidJson,
+                              "Could not open output files: #{e.message}").to_json
```
Comment (richardc) [inline modules/pxp-module-puppet]: This should probably use a new Error, rather than being 'invalid json'

File: modules/pxp-module-puppet [hunk 10]
```diff
@@ -221,12 +228,59 @@ def run(params_and_config)
...
+      end
+
+      begin
+        File.open(output_files["exitcode"], 'w') do |f|
+          f.puts(status)
+        end
+      rescue => e
+        print make_error_result(DEFAULT_EXITCODE, Errors::InvalidJson,
+                                "Could not open exit code file: #{e.message}").to_json
+        exit 5 # this exit code is reserved for problems with opening
```
Comment (richardc) [inline modules/pxp-module-puppet]: This is also not Json was invalid.  Some form of file handling error instead?

**Comment by parisiale** [comment]
@richardc changing the error types would imply changing pxp-module-puppet's interface: https://github.com/puppetlabs/pxp-agent/blob/master/modules/pxp-module-puppet.md.
We don't want to do that here. There's PCP-219 for that. It's ok to use the errors as they are.

**Comment by mruzicka** [comment]
I've squashed the commit into one.
